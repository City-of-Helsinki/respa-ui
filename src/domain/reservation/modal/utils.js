export const getCancelCategories = (data = [], isAdmin, locale) => {
  return data
    .filter(({ reservation_type: reservationType }) => (
      isAdmin ? true : reservationType === 'own'))
    .map(category => ({
      value: category.id,
      label: category.name[locale || 'fi'],
    }));
};
