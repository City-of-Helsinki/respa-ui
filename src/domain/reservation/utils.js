import moment from 'moment';
import find from 'lodash/find';
import sortBy from 'lodash/sortBy';
import clone from 'lodash/clone';
import tail from 'lodash/tail';
import last from 'lodash/last';

export const combineReservations = (reservations) => {
  if (!reservations || !reservations.length) {
    return [];
  }

  const sorted = sortBy(reservations, 'begin');
  const initialValue = [clone(sorted[0])];

  return tail(sorted).reduce((previous, current) => {
    if (current.begin === last(previous).end) {
      last(previous).end = current.end;
    } else {
      previous.push(clone(current));
    }
    return previous;
  }, initialValue);
};

export const getNextAvailableTime = (reservations, fromMoment = moment()) => {
  const combinedReservations = combineReservations(reservations);
  if (!combinedReservations.length || fromMoment < moment(combinedReservations[0].begin)) {
    return fromMoment;
  }

  const ongoingReservation = find(combinedReservations, reservation => (
    moment(reservation.begin) <= fromMoment && fromMoment < moment(reservation.end)
  ));

  return ongoingReservation ? moment(ongoingReservation.end) : fromMoment;
};

export const getCurrentReservation = (reservations) => {
  const now = moment();
  return find(reservations, (reservation) => {
    return moment(reservation.begin) < now && now < moment(reservation.end);
  });
};
