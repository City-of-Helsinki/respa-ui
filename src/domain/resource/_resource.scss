@import './card/resourceCard';
@import './card/slider/resourceCardSlider';
