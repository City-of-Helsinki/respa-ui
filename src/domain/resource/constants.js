export const resourceProductTypes = Object.freeze({
  RENT: 'rent',
  EXTRA: 'extra',
});

export const resourcePriceTypes = Object.freeze({
  HOURLY: 'hourly',
  DAILY: 'daily',
  WEEKLY: 'weekly',
  FIXED: 'fixed',
});

export const productPriceType = Object.freeze({
  PER_PERIOD: 'per_period',
  FIXED: 'fixed',
});

export const resourcePeriodLengthType = Object.freeze({
  WITHIN_DAY: 'within_day',
  WHOLE_DAY: 'whole_day',
  OVERNIGHT: 'over_night',
});
