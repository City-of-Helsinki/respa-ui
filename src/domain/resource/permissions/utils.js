import difference from 'lodash/difference';

import { resourceRoles, resourcePermissionTypes, resourcePermissionsByRole } from './constants';

// Returns the role with the most permissions or null.
export function roleMapper(isAdmin, isManager, isViewer) {
  if (isAdmin) {
    return resourceRoles.UNIT_ADMINISTRATOR;
  }

  if (isManager) {
    return resourceRoles.UNIT_MANAGER;
  }

  if (isViewer) {
    return resourceRoles.UNIT_VIEWER;
  }

  return null;
}

const permissionOptions = Object.values(resourcePermissionTypes);
const allValuesInArray = (subset, superset) => difference(subset, superset).length === 0;

export function hasPermissionForResource(resourceRole, requiredPermissions) {
  // undefined roles don't receive any permissions implicitly
  if (resourceRole === undefined || resourceRole === null) {
    return false;
  }

  if (!Object.values(resourceRoles).includes(resourceRole)) {
    throw Error(`resourceRole needs to be one of: ${Object.values(resourceRoles).join(', ')}.`);
  }

  if (!Array.isArray(requiredPermissions) && typeof requiredPermissions !== 'string') {
    throw Error('requiredPermissions has to be either a string or an array.');
  }

  if (
    Array.isArray(requiredPermissions)
    && !allValuesInArray(requiredPermissions, permissionOptions)
  ) {
    const unknownPermissions = difference(requiredPermissions, permissionOptions);

    // eslint-disable-next-line no-console
    console.warn([
      `Permission asked with unknown permission: ${unknownPermissions.join(', ')}.`,
      `Permission needs to be on of: ${permissionOptions.join(', ')}.`,
    ].join('\n'));
  }

  if (typeof requiredPermissions === 'string' && !permissionOptions.includes(requiredPermissions)) {
    // eslint-disable-next-line no-console
    console.warn([
      `Permission asked with unknown permission: ${requiredPermissions}.`,
      `Permission needs to be on of: ${permissionOptions.join(', ')}.`,
    ].join('\n'));
  }

  const rolePermissions = resourcePermissionsByRole[resourceRole];

  if (Array.isArray(requiredPermissions)) {
    return allValuesInArray(requiredPermissions, rolePermissions);
  }

  return rolePermissions.includes(requiredPermissions);
}
