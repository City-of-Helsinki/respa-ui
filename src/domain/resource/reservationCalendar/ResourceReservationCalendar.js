import * as React from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import FullCalendar from '@fullcalendar/react';
import timeGridPlugin from '@fullcalendar/timegrid';
import momentTimezonePlugin from '@fullcalendar/moment-timezone';
import interactionPlugin from '@fullcalendar/interaction';
import enLocale from '@fullcalendar/core/locales/en-gb';
import svLocale from '@fullcalendar/core/locales/sv';
import fiLocale from '@fullcalendar/core/locales/fi';
import moment from 'moment';
import get from 'lodash/get';
import classNames from 'classnames';
import Button from 'react-bootstrap/lib/Button';

import constants from '../../../../app/constants/AppConstants';
import * as resourceUtils from '../utils';
import injectT from '../../../../app/i18n/injectT';

class ResourceReservationCalendar extends React.Component {
  calendarRef = React.createRef();

  static propTypes = {
    date: PropTypes.string,
    intl: intlShape,
    resource: PropTypes.object.isRequired,
    onDateChange: PropTypes.func.isRequired,
    t: PropTypes.func.isRequired,
  };

  state = {
    selected: null,
    viewType: 'timeGridWeek',
  };

  componentDidUpdate(prevProps) {
    const { date } = this.props;

    if (date !== prevProps.date) {
      const calendarApi = this.calendarRef.current.getApi();
      calendarApi.gotoDate(date);
    }
  }

  getCalendarOptions = () => {
    const { intl } = this.props;

    return {
      header: {
        left: 'prev,next today',
        center: 'title',
        right: 'timeGridDay,timeGridWeek'
      },
      height: 'auto',
      editable: true,
      eventConstraint: 'businessHours',
      eventOverlap: false,
      firstDay: 1,
      locale: intl.locale,
      locales: [enLocale, svLocale, fiLocale],
      nowIndicator: true,
      plugins: [timeGridPlugin, momentTimezonePlugin, interactionPlugin],
      // selectable: true,
      selectOverlap: false,
      selectConstraint: 'businessHours',
      slotLabelFormat: {
        hour: 'numeric',
        minute: '2-digit',
        omitZeroMinute: false,
        meridiem: 'short'
      },
      unselectAuto: false,
    };
  };

  getEvents = () => {
    const { resource, date } = this.props;
    const { selected } = this.state;

    const getClassNames = (reservation) => {
      const isOwn = get(reservation, 'is_own', false);
      return classNames('app-ResourceReservationCalendar__event', {
        'app-ResourceReservationCalendar__event--reserved': !isOwn,
      });
    };

    // Add the resources reservations as normal FullCalendar events.
    const events = get(resource, 'reservations', []).map(reservation => ({
      classNames: [getClassNames(reservation)],
      editable: false,
      id: reservation.id,
      start: moment(reservation.begin).toDate(),
      end: moment(reservation.end).toDate(),
    }));

    // Check resources reservation rules and disable days if needed.
    const now = moment();
    const momentDate = moment(date)
      .startOf('week');

    for (let i = 0; i < 7; i++) {
      if (
        momentDate.isBefore(now, 'date')
        || !resourceUtils.isDateReservable(resource, momentDate.format(constants.DATE_FORMAT))
      ) {
        events.push({
          allDay: true,
          classNames: [
            'app-ResourceReservationCalendar__backgroundEvent',
            'app-ResourceReservationCalendar__restrictedDate',
          ],
          id: momentDate.format(constants.DATE_FORMAT),
          start: momentDate.toDate(),
          end: momentDate.toDate(),
          rendering: 'background',
        });
      }

      momentDate.add(1, 'day');
    }

    // Add the selected time range into calendar as an event.
    if (selected) {
      events.push({
        classNames: [
          'app-ResourceReservationCalendar__event',
          'app-ResourceReservationCalendar__newReservation',
        ],
        editable: true,
        id: 'newReservation',
        ...selected,
      });
    }

    return events;
  };

  onDatesRender = (info) => {
    const {
      date,
      onDateChange,
    } = this.props;

    const momentDate = moment(date);
    const activeStart = moment(info.view.activeStart);

    // For some weird reason the activeEnd date is always last day of the day/week view + 1
    // (e.g. for a week view it's always the next weeks monday);
    const activeEnd = moment(info.view.activeEnd).subtract(1);

    if (momentDate.isBefore(activeStart, 'day') || momentDate.isAfter(activeEnd, 'day')) {
      onDateChange(activeStart.format(constants.DATE_FORMAT));
    }

    const {
      viewType,
    } = this.state;

    if (viewType !== info.view.type) {
      this.setState({
        viewType: info.view.type,
      });
    }
  };

  onDateClick = (dateClickInfo) => {
    const { resource } = this.props;
    const slotSize = resourceUtils.getSlotSizeInMinutes(resource);

    const start = moment(dateClickInfo.date);
    const end = start.clone().add(slotSize, 'minutes');

    if (this.isEventValid(start.toDate(), end.toDate())) {
      const calendarApi = this.calendarRef.current.getApi();
      calendarApi.select({
        start: start.toDate(),
        end: end.toDate(),
      });
    }
  };

  onSelect = (selectionInfo) => {
    this.setState({
      selected: {
        start: selectionInfo.start,
        end: selectionInfo.end,
      },
    });

    // Hide the FullCalendar selection widget/indicator.
    const calendarApi = this.calendarRef.current.getApi();
    calendarApi.unselect();
  };

  onEventAllow = (dropInfo) => {
    return this.isEventValid(dropInfo.start, dropInfo.end);
  };

  onSelectAllow = (selectInfo) => {
    return this.isEventValid(selectInfo.start, selectInfo.end);
  };

  /**
   * isEventValid();
   * @param start {Date}
   * @param end {Date}
   * @returns {boolean}
   */
  isEventValid = (start, end) => {
    const { resource } = this.props;

    const now = moment();
    const startMoment = moment(start);
    const endMoment = moment(end);

    // Reservation cannot be longer than the resources max period if max period is set.
    const maxPeriod = get(resource, 'max_period', null);
    if (maxPeriod) {
      const maxPeriodDuration = moment.duration(maxPeriod);
      const maxDuration = maxPeriodDuration.hours() * 60 + maxPeriodDuration.minutes();

      if (endMoment.diff(startMoment, 'minutes') > maxDuration) {
        return false;
      }
    }

    if (!resourceUtils.isDateReservable(resource, startMoment.format(constants.DATE_FORMAT))
      || !resourceUtils.isDateReservable(resource, endMoment.format(constants.DATE_FORMAT))) {
      return false;
    }

    // Check if the given event times are inside opening hours.
    const dateString = startMoment.format(constants.DATE_FORMAT);
    const openingHours = resourceUtils.getOpeningHours(resource, dateString);
    const opens = moment(openingHours.opens);
    const closes = moment(openingHours.closes);

    if (startMoment.isBefore(opens) || endMoment.isAfter(closes)) {
      return false;
    }

    // Prevent selecting times from past.
    return startMoment.isAfter(now);
  };

  onEventDrop = (eventDropInfo) => {
    const { event } = eventDropInfo;

    this.setState({
      selected: {
        start: event.start,
        end: event.end,
      },
    });
  };

  onEventResize = (eventResizeInfo) => {
    const { event } = eventResizeInfo;

    this.setState({
      selected: {
        start: event.start,
        end: event.end,
      },
    });
  };

  onReserveButtonClick = () => {

  };

  getDurationText = () => {
    const { selected } = this.state;
    const start = moment(selected.start);
    const end = moment(selected.end);
    const duration = moment.duration(end.diff(start));
    const days = duration.days();
    const hours = duration.hours();
    const minutes = duration.minutes();

    let text = '';
    if (days) {
      text = `${days}d`;
    }

    if (hours) {
      text += `${hours}h`;
    }

    if (minutes) {
      text += `${minutes}min`;
    }

    return text;
  };

  getSelectedDateText = () => {
    const { t } = this.props;
    const { selected } = this.state;

    if (selected) {
      const start = moment(selected.start);
      const end = moment(selected.end);

      return t('ResourceReservationCalendar.selectedDateValue', {
        date: start.format('dd D.M.Y'),
        start: start.format('HH:mm'),
        end: end.format('HH:mm'),
        duration: this.getDurationText(),
      });
    }

    return '';
  };

  render() {
    const {
      date,
      resource,
      t,
    } = this.props;

    const {
      viewType,
      selected,
    } = this.state;

    return (
      <div className="app-ResourceReservationCalendar">
        <p>Slot size: {resource.slot_size}</p>
        <p>Min & max period: {resource.min_period} - {resource.max_period}</p>

        <FullCalendar
          allDaySlot={false}
          businessHours={resourceUtils.getFullCalendarBusinessHours(resource, date)}
          dateClick={this.onDateClick}
          datesRender={this.onDatesRender}
          defaultDate={date}
          eventAllow={this.onEventAllow}
          eventDrop={this.onEventDrop}
          eventResize={this.onEventResize}
          events={this.getEvents()}
          ref={this.calendarRef}
          select={this.onSelect}
          selectAllow={this.onSelectAllow}
          slotDuration={resourceUtils.getFullCalendarSlotDuration(resource, date, viewType)}
          slotLabelInterval={resourceUtils.getFullCalendarSlotLabelInterval(resource)}
          {...this.getCalendarOptions()}
          maxTime={resourceUtils.getFullCalendarMaxTime(resource, date, viewType)}
          minTime={resourceUtils.getFullCalendarMinTime(resource, date, viewType)}
        />

        {selected && (
          <div className="app-ResourceReservationCalendar__selectedInfo">
            <div className="app-ResourceReservationCalendar__selectedDate">
              <strong className="app-ResourceReservationCalendar__selectedDateLabel">
                {t('ResourceReservationCalendar.selectedDateLabel')}
              </strong>
              {' '}
              <span className="app-ResourceReservationCalendar__selectedDateValue">
                {this.getSelectedDateText()}
              </span>
            </div>
            <Button
              bsStyle="primary"
              className="app-ResourceReservationCalendar__reserveButton"
              onClick={() => this.onReserveButtonClick()}
            >
              {t('ResourceReservationCalendar.reserveButton')}
            </Button>
          </div>
        )}
      </div>
    );
  }
}

export default injectT(injectIntl(ResourceReservationCalendar));
