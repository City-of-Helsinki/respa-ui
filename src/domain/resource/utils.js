import round from 'lodash/round';
import filter from 'lodash/filter';
import findIndex from 'lodash/findIndex';
import get from 'lodash/get';
import find from 'lodash/find';
import moment from 'moment';

import * as urlUtils from '../../common/url/utils';
import * as dataUtils from '../../common/data/utils';
import * as reservationUtils from '../reservation/utils';
import constants from '../../../app/constants/AppConstants';
import forEach from "lodash/forEach";


/**
 * Getter for resource page link.
 * Returns a resource page url with the given query params.
 * @param resource
 * @param query
 * @returns {string}
 */
export const getResourcePageLink = (resource, query) => {
  return urlUtils.getLinkString(`/resources/${resource.id}`, query);
};

/**
 * Getter for formatted unit address string.
 * @param unit
 * @param locale
 * @returns {string}
 */
export const getUnitAddress = (unit, locale) => {
  let postal;
  const streetAddress = dataUtils.getLocalizedFieldValue(unit.street_address, locale);
  if (unit && unit.address_zip) {
    postal = unit.address_zip;
  }
  if (unit && unit.municipality) {
    postal = postal ? `${postal} ${unit.municipality}` : unit.municipality;
  }

  return `${streetAddress}${postal ? ` ${postal}` : ''}`;
};

/**
 * Getter for formatted resource distance string.
 * @param resource
 * @returns {string}
 */
export const getResourceDistance = (resource) => {
  const km = resource.distance / 1000;
  return km ? `${round(km, km < 10 ? 1 : null)} km` : '';
};

/**
 * Getter for price string used in resource cards.
 * @param resource
 * @param t
 * @returns {string|*}
 */
export const getPrice = (resource, t) => {
  const minPricePerHour = resource.min_price_per_hour;
  const maxPricePerHour = resource.max_price_per_hour;

  if (!(minPricePerHour || maxPricePerHour)) {
    return t('ResourceIcons.free');
  }

  if (minPricePerHour && maxPricePerHour && minPricePerHour !== maxPricePerHour) {
    return `${Number(minPricePerHour)} - ${Number(maxPricePerHour)} €/h`;
  }

  const priceString = maxPricePerHour || minPricePerHour;
  const price = priceString !== 0 ? Number(priceString) : 0;
  if (price === 0) {
    return t('ResourceIcons.free');
  }

  return price ? `${price} €/h` : null;
};

export const isFree = (resource) => {
  const minPricePerHour = resource.min_price_per_hour;
  const maxPricePerHour = resource.max_price_per_hour;

  if (!(minPricePerHour || maxPricePerHour)) {
    return true;
  }

  if (minPricePerHour && maxPricePerHour && minPricePerHour !== maxPricePerHour) {
    return false;
  }

  const priceString = maxPricePerHour || minPricePerHour;
  const price = priceString !== 0 ? Number(priceString) : 0;

  return price === 0;
};

export const getOpeningHours = (resource, date = null) => {
  if (!resource) {
    return null;
  }

  const openingHours = get(resource, 'opening_hours', []);
  if (!openingHours.length) {
    return null;
  }

  const index = date ? findIndex(openingHours, item => item.date === date) : 0;
  return {
    closes: get(openingHours, `[${index}].closes`),
    opens: get(openingHours, `[${index}].opens`),
  };
};

const reservingIsRestricted = (resource, date) => {
  if (!date) {
    return false;
  }

  const isAdmin = get(resource, 'user_permissions.is_admin', false);
  const isLimited = resource.reservable_before && moment(resource.reservable_before).isBefore(moment(date), 'day');

  return isLimited && !isAdmin;
};

export const getOpenReservations = (resource) => {
  return filter(resource.reservations, (reservation) => {
    return reservation.state !== 'cancelled' && reservation.state !== 'denied';
  });
};

export const getAvailabilityDataForNow = (resource, date) => {
  const openingHours = getOpeningHours(resource, date);
  const reservations = getOpenReservations(resource);

  if (!openingHours || !openingHours.closes || !openingHours.opens) {
    return { status: 'closed', bsStyle: 'danger' };
  }

  const nowMoment = moment();
  const opensMoment = moment(openingHours.opens);
  const closesMoment = moment(openingHours.closes);
  const beginMoment = nowMoment > opensMoment ? nowMoment : opensMoment;

  if (nowMoment > closesMoment) {
    return { status: 'closed', bsStyle: 'danger' };
  }

  const currentReservation = reservationUtils.getCurrentReservation(reservations);
  if (currentReservation || nowMoment < opensMoment) {
    const nextAvailableTime = reservationUtils.getNextAvailableTime(reservations, beginMoment);
    if (nextAvailableTime < closesMoment) {
      return {
        status: 'availableAt',
        bsStyle: 'danger',
        values: { time: nextAvailableTime.format(constants.TIME_FORMAT) },
      };
    }
    return { status: 'reserved', bsStyle: 'danger' };
  }

  return { status: 'available', bsStyle: 'success' };
};

export const getAvailabilityDataForWholeDay = (resource, date) => {
  const openingHours = getOpeningHours(resource, date);

  if (!openingHours || !openingHours.closes || !openingHours.opens) {
    return { status: 'closed', bsStyle: 'danger' };
  }

  if (reservingIsRestricted(resource, date)) {
    return { status: 'reservingRestricted', bsStyle: 'danger' };
  }

  const reservations = getOpenReservations(resource);

  const opensMoment = moment(openingHours.opens);
  const closesMoment = moment(openingHours.closes);
  let total = closesMoment - opensMoment;

  forEach(reservations, (reservation) => {
    const resBeginMoment = moment(reservation.begin);
    if (!resBeginMoment.isSame(opensMoment, 'd')) {
      return;
    }

    const resEndMoment = moment(reservation.end);
    total -= resEndMoment;
    total += resBeginMoment;
  });

  const asHours = moment.duration(total).asHours();
  const rounded = Math.ceil(asHours * 2) / 2;

  if (rounded === 0) {
    return { status: 'reserved', bsStyle: 'danger' };
  }

  return {
    status: 'availableTime',
    bsStyle: 'success',
    values: { hours: rounded },
  };
};
