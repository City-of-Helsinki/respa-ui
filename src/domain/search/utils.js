import get from 'lodash/get';
import range from 'lodash/range';
import capitalize from 'lodash/capitalize';
import sortBy from 'lodash/sortBy';

import constants from '../../../app/constants/AppConstants';

// TODO: Change to use query-string module.
export const getFiltersFromUrl = (location) => {
  const query = new URLSearchParams(location.search);
  const filters = {};

  query.forEach((value, key) => {
    if (constants.SUPPORTED_SEARCH_FILTERS[key] !== undefined) {
      filters[key] = value;
    }
  });

  return filters;
};

// TODO: Change to use query-string module.
export const getSearchFromFilters = (filters) => {
  const query = new URLSearchParams(filters);
  return query.toString();
};

export const getUnitOptions = (units, locale) => {
  const options = units.map(unit => ({
    value: unit.id,
    label: get(unit, `name[${locale}]`, ''),
  }));

  return sortBy(options, 'label');
};

export const getPurposeOptions = (purposes, locale) => {
  const options = purposes
    .filter(purpose => purpose.parent === null)
    .map(unit => ({
      value: unit.id,
      label: get(unit, `name[${locale}]`, ''),
    }));

  return sortBy(options, 'label');
};

/**
 * Getter for people capacity options.
 * @returns {{label: number, value: number}[]}
 */
export const getPeopleCapacityOptions = () => {
  return [
    ...range(1, 10),
    ...range(10, 35, 5),
    ...range(40, 110, 10),
  ].map(number => ({ label: number, value: number }));
};

/**
 * Getter for municipality options.
 * @returns {{label: string, value: string}[]}
 */
export const getMunicipalityOptions = () => {
  let municipalities = constants.DEFAULT_MUNICIPALITY_OPTIONS;

  if (Array.isArray(SETTINGS.CUSTOM_MUNICIPALITY_OPTIONS)
    && SETTINGS.CUSTOM_MUNICIPALITY_OPTIONS.length) {
    municipalities = SETTINGS.CUSTOM_MUNICIPALITY_OPTIONS;
  }

  return municipalities.map((municipality) => {
    const municipalityStr = typeof municipality === 'string' ? municipality : municipality.toString();

    return {
      value: municipalityStr.toLowerCase(),
      label: capitalize(municipalityStr),
    };
  });
};
