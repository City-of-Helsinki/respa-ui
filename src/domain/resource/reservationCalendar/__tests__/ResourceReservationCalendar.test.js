import React from 'react';
import toJSON from 'enzyme-to-json';

import { shallowWithIntl, globalDateMock } from '../../../../../app/utils/testUtils';
import ResourceReservationCalendar from '../ResourceReservationCalendar';
import resource from '../../../../common/data/fixtures/resource';

describe('ResourceReservationCalendar', () => {
  test('renders correctly', () => {
    globalDateMock();
    const props = {
      resource: resource.build(),
      date: '2019-08-15',
      onDateChange: jest.fn(),
    };

    const wrapper = shallowWithIntl(
      <ResourceReservationCalendar {...props} />
    );

    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
