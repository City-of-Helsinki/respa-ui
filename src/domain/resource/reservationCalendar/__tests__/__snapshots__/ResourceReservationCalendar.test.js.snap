// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceReservationCalendar renders correctly 1`] = `
<div
  className="app-ResourceReservationCalendar"
>
  <FullCalendar
    allDaySlot={false}
    businessHours={
      Array [
        Object {
          "daysOfWeek": Array [
            1,
          ],
          "endTime": "00:00",
          "startTime": "00:00",
        },
        Object {
          "daysOfWeek": Array [
            2,
          ],
          "endTime": "00:00",
          "startTime": "00:00",
        },
        Object {
          "daysOfWeek": Array [
            3,
          ],
          "endTime": "00:00",
          "startTime": "00:00",
        },
        Object {
          "daysOfWeek": Array [
            4,
          ],
          "endTime": "00:00",
          "startTime": "00:00",
        },
        Object {
          "daysOfWeek": Array [
            5,
          ],
          "endTime": "00:00",
          "startTime": "00:00",
        },
        Object {
          "daysOfWeek": Array [
            6,
          ],
          "endTime": "00:00",
          "startTime": "00:00",
        },
        Object {
          "daysOfWeek": Array [
            0,
          ],
          "endTime": "00:00",
          "startTime": "00:00",
        },
      ]
    }
    dateClick={[Function]}
    datesRender={[Function]}
    defaultDate="2019-08-15"
    editable={true}
    eventAllow={[Function]}
    eventConstraint="businessHours"
    eventDrop={[Function]}
    eventOverlap={false}
    eventRender={[Function]}
    eventResize={[Function]}
    events={
      Array [
        Object {
          "allDay": true,
          "classNames": Array [
            "app-ResourceReservationCalendar__backgroundEvent",
            "app-ResourceReservationCalendar__restrictedDate",
          ],
          "end": 2019-08-11T21:00:00.000Z,
          "id": "2019-08-12",
          "rendering": "background",
          "start": 2019-08-11T21:00:00.000Z,
        },
        Object {
          "allDay": true,
          "classNames": Array [
            "app-ResourceReservationCalendar__backgroundEvent",
            "app-ResourceReservationCalendar__restrictedDate",
          ],
          "end": 2019-08-12T21:00:00.000Z,
          "id": "2019-08-13",
          "rendering": "background",
          "start": 2019-08-12T21:00:00.000Z,
        },
        Object {
          "allDay": true,
          "classNames": Array [
            "app-ResourceReservationCalendar__backgroundEvent",
            "app-ResourceReservationCalendar__restrictedDate",
          ],
          "end": 2019-08-13T21:00:00.000Z,
          "id": "2019-08-14",
          "rendering": "background",
          "start": 2019-08-13T21:00:00.000Z,
        },
        Object {
          "allDay": true,
          "classNames": Array [
            "app-ResourceReservationCalendar__backgroundEvent",
            "app-ResourceReservationCalendar__restrictedDate",
          ],
          "end": 2019-08-14T21:00:00.000Z,
          "id": "2019-08-15",
          "rendering": "background",
          "start": 2019-08-14T21:00:00.000Z,
        },
        Object {
          "allDay": true,
          "classNames": Array [
            "app-ResourceReservationCalendar__backgroundEvent",
            "app-ResourceReservationCalendar__restrictedDate",
          ],
          "end": 2019-08-15T21:00:00.000Z,
          "id": "2019-08-16",
          "rendering": "background",
          "start": 2019-08-15T21:00:00.000Z,
        },
        Object {
          "allDay": true,
          "classNames": Array [
            "app-ResourceReservationCalendar__backgroundEvent",
            "app-ResourceReservationCalendar__restrictedDate",
          ],
          "end": 2019-08-16T21:00:00.000Z,
          "id": "2019-08-17",
          "rendering": "background",
          "start": 2019-08-16T21:00:00.000Z,
        },
        Object {
          "allDay": true,
          "classNames": Array [
            "app-ResourceReservationCalendar__backgroundEvent",
            "app-ResourceReservationCalendar__restrictedDate",
          ],
          "end": 2019-08-17T21:00:00.000Z,
          "id": "2019-08-18",
          "rendering": "background",
          "start": 2019-08-17T21:00:00.000Z,
        },
      ]
    }
    firstDay={1}
    header={
      Object {
        "center": "title",
        "left": "prev,next today",
        "right": "timeGridDay,timeGridWeek",
      }
    }
    height="auto"
    locale="en"
    locales={
      Array [
        Object {
          "code": "en-gb",
          "week": Object {
            "dow": 1,
            "doy": 4,
          },
        },
        Object {
          "allDayText": "Heldag",
          "buttonText": Object {
            "day": "Dag",
            "list": "Program",
            "month": "Månad",
            "next": "Nästa",
            "prev": "Förra",
            "today": "Idag",
            "week": "Vecka",
          },
          "code": "sv",
          "eventLimitText": "till",
          "noEventsMessage": "Inga händelser att visa",
          "week": Object {
            "dow": 1,
            "doy": 4,
          },
          "weekLabel": "v.",
        },
        Object {
          "allDayText": "Koko päivä",
          "buttonText": Object {
            "day": "Päivä",
            "list": "Tapahtumat",
            "month": "Kuukausi",
            "next": "Seuraava",
            "prev": "Edellinen",
            "today": "Tänään",
            "week": "Viikko",
          },
          "code": "fi",
          "eventLimitText": "lisää",
          "noEventsMessage": "Ei näytettäviä tapahtumia",
          "week": Object {
            "dow": 1,
            "doy": 4,
          },
          "weekLabel": "Vk",
        },
      ]
    }
    maxTime="17:00:00"
    minTime="07:00:00"
    nowIndicator={true}
    plugins={
      Array [
        Object {
          "calendarInteractions": Array [],
          "cmdFormatter": undefined,
          "componentInteractions": Array [],
          "datePointTransforms": Array [],
          "dateSelectionTransformers": Array [],
          "dateSpanTransforms": Array [],
          "defaultView": "timeGridWeek",
          "deps": Array [],
          "elementDraggingImpl": undefined,
          "eventDefMutationAppliers": Array [],
          "eventDefParsers": Array [],
          "eventDragMutationMassagers": Array [],
          "eventDropTransformers": Array [],
          "eventResizeJoinTransforms": Array [],
          "eventSourceDefs": Array [],
          "externalDefTransforms": Array [],
          "id": "6",
          "isDraggableTransformers": Array [],
          "isPropsValid": null,
          "namedTimeZonedImpl": undefined,
          "optionChangeHandlers": Object {},
          "recurringTypes": Array [],
          "reducers": Array [],
          "themeClasses": Object {},
          "viewContainerModifiers": Array [],
          "viewPropsTransformers": Array [],
          "views": Object {
            "timeGrid": Object {
              "allDaySlot": true,
              "class": [Function],
              "slotDuration": "00:30:00",
              "slotEventOverlap": true,
            },
            "timeGridDay": Object {
              "duration": Object {
                "days": 1,
              },
              "type": "timeGrid",
            },
            "timeGridWeek": Object {
              "duration": Object {
                "weeks": 1,
              },
              "type": "timeGrid",
            },
          },
        },
        Object {
          "calendarInteractions": Array [],
          "cmdFormatter": undefined,
          "componentInteractions": Array [],
          "datePointTransforms": Array [],
          "dateSelectionTransformers": Array [],
          "dateSpanTransforms": Array [],
          "defaultView": "",
          "deps": Array [],
          "elementDraggingImpl": undefined,
          "eventDefMutationAppliers": Array [],
          "eventDefParsers": Array [],
          "eventDragMutationMassagers": Array [],
          "eventDropTransformers": Array [],
          "eventResizeJoinTransforms": Array [],
          "eventSourceDefs": Array [],
          "externalDefTransforms": Array [],
          "id": "7",
          "isDraggableTransformers": Array [],
          "isPropsValid": null,
          "namedTimeZonedImpl": [Function],
          "optionChangeHandlers": Object {},
          "recurringTypes": Array [],
          "reducers": Array [],
          "themeClasses": Object {},
          "viewContainerModifiers": Array [],
          "viewPropsTransformers": Array [],
          "views": Object {},
        },
        Object {
          "calendarInteractions": Array [
            [Function],
          ],
          "cmdFormatter": undefined,
          "componentInteractions": Array [
            [Function],
            [Function],
            [Function],
            [Function],
          ],
          "datePointTransforms": Array [],
          "dateSelectionTransformers": Array [],
          "dateSpanTransforms": Array [],
          "defaultView": "",
          "deps": Array [],
          "elementDraggingImpl": [Function],
          "eventDefMutationAppliers": Array [],
          "eventDefParsers": Array [],
          "eventDragMutationMassagers": Array [],
          "eventDropTransformers": Array [],
          "eventResizeJoinTransforms": Array [],
          "eventSourceDefs": Array [],
          "externalDefTransforms": Array [],
          "id": "8",
          "isDraggableTransformers": Array [],
          "isPropsValid": null,
          "namedTimeZonedImpl": undefined,
          "optionChangeHandlers": Object {},
          "recurringTypes": Array [],
          "reducers": Array [],
          "themeClasses": Object {},
          "viewContainerModifiers": Array [],
          "viewPropsTransformers": Array [],
          "views": Object {},
        },
      ]
    }
    select={[Function]}
    selectAllow={[Function]}
    selectConstraint="businessHours"
    selectOverlap={false}
    selectable={true}
    slotDuration="00:30:00"
    slotLabelFormat={
      Object {
        "hour": "numeric",
        "meridiem": "short",
        "minute": "2-digit",
        "omitZeroMinute": false,
      }
    }
    slotLabelInterval="01:00:00"
    unselectAuto={false}
  />
</div>
`;
